arroyo/processing/processor.py

```diff
@@ -207,6 +210,19 @@ def run(self) -> None:
             logger.info("Processor terminated")
             raise
 
+    def _handle_invalid_message(self, exc: InvalidMessage) -> None:
+        logger.exception(exc)
+        if self.__dlq_policy:
+            invalid_message = self.__buffered_messages.pop(exc.partition, exc.offset)
```

Comment:
Is there actually any real need to allow `InvalidMessage` to be received out of order? I can't really think of a scenario where that is needed right now. These assumptions exist for all messages in consumers, and I'm not sure there's benefit in relaxing them for DLQ messages only. I think it is a violation of the kinds of guarantees we want to provide, and there's no reason to not be strict. Yes it's true it could change one day, but I would propose to tackle it all at once and not have a sort of randomly different constraint for invalid messages today.

Reply:
On second thoughts, while I think it' still unlikely via Arroyo's strategies, it may be possible in user defined strategies which don't do the right thing. Not sure how much we want to guard against this though.

Reply:
> Not sure how much we want to guard against this though.

I think the risk/benefit trade off makes it worth accepting exceptions out of order in this case:
- If the product does the wrong thing, they will notice when they are trying to recover from an error condition (invalid message to be sent to a DLQ). It will be a very easy case to miss and will make the consumer crash exactly while going through the code that prevents it from crashing. Not the greatest place to have a hidden assumption.
- While we try to process messages in order, it is fairly easy for a client to write a custom strategy that does not do that without violating any constraints in the API. There are also good use cases for this (think about replay segments). So it is bound to happen sooner than later, not just hypothetical.

So overall, error management logic (the DLQ) should be very robust as it is supposed to be tested rarely in prod. You do not want to have it fail because of a wrong assumption made by the author of a consumer when it is needed most. The complexity to support out of order exceptions does not seem sensibly higher.

Reply:
I changed this to a warning log and made it unconfigurable. I think people should still fix their strategies. It's just not worth it dropping their DLQ items over.